packages/seedless-onboarding-controller/src/SeedlessOnboardingController.ts

```diff
@@ -0,0 +1,829 @@
+   */
+  async #withPersistedSeedPhraseBackupsState<
+    Result extends Uint8Array | Uint8Array[],
+  >(callback: () => Promise<Result>): Promise<Result> {
+    try {
+      const backedUpSeedPhrases = await callback();
+      let backedUpHashB64Strings: string[] = [];
+
+      if (Array.isArray(backedUpSeedPhrases)) {
+        backedUpHashB64Strings = backedUpSeedPhrases.map((seedPhrase) =>
+          keccak256AndHexify(seedPhrase),
+        );
+      } else {
+        backedUpHashB64Strings = [keccak256AndHexify(backedUpSeedPhrases)];
+      }
+
+      const existingBackedUpHashes = this.state.backupHashes;
+      const uniqueHashesSet = new Set([
+        ...existingBackedUpHashes,
+        ...backedUpHashB64Strings,
+      ]);
+
+      this.update((state) => {
+        state.backupHashes = Array.from(uniqueHashesSet);
+      });
+
+      return backedUpSeedPhrases;
+    } catch (error) {
+      log('Error persisting seed phrase backups', error);
+      throw error;
+    }
+  }
+
+  /**
+   * Create a new vault with the given authentication data.
+   *
+   * Serialize the authentication and key data which will be stored in the vault.
+   *
+   * @param params - The parameters for creating a new vault.
+   * @param params.password - The password to encrypt the vault.
+   * @param params.rawToprfEncryptionKey - The encryption key to encrypt the vault.
+   * @param params.rawToprfAuthKeyPair - The authentication key pair to encrypt the vault.
+   */
+  async #createNewVaultWithAuthData({
+    password,
+    rawToprfEncryptionKey,
+    rawToprfAuthKeyPair,
+  }: {
+    password: string;
+    rawToprfEncryptionKey: Uint8Array;
+    rawToprfAuthKeyPair: KeyPair;
+  }): Promise<void> {
+    const { nodeAuthTokens } = this.state;
+    this.#assertIsValidNodeAuthTokens(nodeAuthTokens);
+
+    const { toprfEncryptionKey, toprfAuthKeyPair } = this.#serializeKeyData(
+      rawToprfEncryptionKey,
+      rawToprfAuthKeyPair,
+    );
+
+    const serializedVaultData = JSON.stringify({
+      authTokens: nodeAuthTokens,
+      toprfEncryptionKey,
+      toprfAuthKeyPair,
+    });
+
+    await this.#updateVault({
+      password,
+      serializedVaultData,
+    });
+  }
+
+  /**
+   * Encrypt and update the vault with the given authentication data.
+   *
+   * @param params - The parameters for updating the vault.
+   * @param params.password - The password to encrypt the vault.
+   * @param params.serializedVaultData - The serialized authentication data to update the vault with.
+   * @returns A promise that resolves to the updated vault.
+   */
+  async #updateVault({
+    password,
+    serializedVaultData,
+  }: {
+    password: string;
+    serializedVaultData: string;
+  }): Promise<void> {
+    await this.#withVaultLock(async () => {
+      assertIsValidPassword(password);
+
+      const updatedState: Partial<SeedlessOnboardingControllerState> = {};
+
+      updatedState.vault = await this.#vaultEncryptor.encrypt(
```

Comment:
Note that encryption using password is also slow because of slow key derivation function. Same comments apply as for decryption.